### ISSUE TITLE
Race condition in xmlrpc

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3152]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#3152|
|Type|defect|
|Created|2008-04-03 05:26:42Z|
|Branch|https://github.com/twisted/twisted/tree/xmlrpc-race-3152|

t.web.xmlrpc has code like this:

```python
class _QueryFactory(protocol.ClientFactory):

# ...

    def parseResponse(self, contents):
        if not self.deferred:
            return
        try:
            response = xmlrpclib.loads(contents)
        except:
            self.deferred.errback(failure.Failure())
            self.deferred = None
        else:
            self.deferred.callback(response[0][0])
            self.deferred = None

    def clientConnectionLost(self, _, reason):
        if self.deferred is not None:
            self.deferred.errback(reason)
            self.deferred = None
```

If the client disconnects while the callback is being processed, `self.deferred` is fired twice.

Will try to repro later. Am focused on identfying issues for our Twisted upgrade.

Attachments:

* [xmlrpc-patch-no-tests.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/887/887705229a4a68507120316cfc13273e1c3777e1/29e7d14a8588f0bfc21beebb0abd17c639c2c67c.diff) (971 bytes) - added by jml on 2008-04-03 05:31:08Z - Patch to remove race condition. No tests.

<details><summary>Searchable metadata</summary>

```
trac-id__3152 3152
type__defect defect
reporter__jml jml
priority__highest highest
milestone__ 
branch__branches_xmlrpc_race_3152 branches/xmlrpc-race-3152
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1207200402000000 1207200402000000
changetime__1207438747000000 1207438747000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

Still haven't minimally reproduced the race or found out what change to Twisted triggers it.

Note that I was getting this reliably in my code before I patched Twisted.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

This patch sponsored by Canonical Ltd.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

(In [23193]) Branching to 'xmlrpc-race-3152'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml removed owner|
|-|-|

This is ready for review.

I didn't test the changes to `clientConnectionLost` -- I couldn't think of any test that wouldn't be grossly artificial. If it needs to be tested before landing, then I'd appreciate being told what sort of test would be appropriate.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @jml|
|-|-|

What about something like that: 
```
    def test_connectionLostCallbackSafety(self):
        d = self.queryFactory.deferred
        d.addErrback(lambda x: self.queryFactory.parseResponse(self.goodContents))
        self.queryFactory.clientConnectionLost(None, self.reason)
        return d
```

I'm not sure it would happen in real life, but it triggers the code.

Also, do you feel like fixing badStatus too?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml set owner to @therve|
|-|-|

So, that's what I meant by "I couldn't think of a test that would be grossly artificial". You'll notice that `parseResponse` doesn't check for `self.deferred is None`, because it assumes (correctly) that it won't be called before `clientConnectionLost`.

If I added the test you suggest, I'd have to add code to the class that's used only for tests.

I thought I did fix `badStatus`. Oh, I did in a different branch. One moment.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @jml|
|-|-|

Great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml set status to `closed`|
|-|-|

(In [23210]) Fix race condition in XMLRPC code.

 * Fixes [#3152](https://github.com/twisted/twisted/issues/3152)
 * Author: jml (sponsored by Canonical Ltd)
 * Reviewer: therve

`_QueryFactory` would try to call a Deferred twice if `clientConnectionLost`
was called while the callbacks for `badStatus` or `parseResponse` were being
processed.

This commit fixes this race using the now idiomatic:
```
    deferred, self.deferred = self.deferred, None
    deferred.callback(foo)
```

It also tests the changed code.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
